File: internal/rss/rss.go
package rss

import (
	"context"
	"encoding/xml"
	"html"
	"fmt"
	"io"
	"net/http"
)

type RSSFeed struct {
	Channel struct {
		Title       string    `xml:"title"`
		Link        string    `xml:"link"`
		Description string    `xml:"description"`
		Item        []RSSItem `xml:"item"`
	} `xml:"channel"`
}

type RSSItem struct {
	Title       string `xml:"title"`
	Link        string `xml:"link"`
	Description string `xml:"description"`
	PubDate     string `xml:"pubDate"`
}

func FetchFeed(ctx context.Context, feedURL string) (*RSSFeed, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, feedURL, nil)
	if err != nil {
		return nil, fmt.Errorf("request instantiation error: %w", err)
	}

	req.Header.Set("User-Agent", "gator")

	res, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("network error: %w", err)
	}
	defer res.Body.Close()

	feed := &RSSFeed{}
	body, err := io.ReadAll(res.Body)
	if err = xml.Unmarshal(body, feed); err != nil {
		return nil, fmt.Errorf("xml unmarshalling error: %w", err)
	}

	feed.Channel.Title = html.UnescapeString(feed.Channel.Title)
	feed.Channel.Description = html.UnescapeString(feed.Channel.Description)
	for _, item := range feed.Channel.Item {
		item.Title = html.UnescapeString(item.Title)
		item.Description = html.UnescapeString(item.Description)
	}

	return feed, nil
}

File: src/rss.ts
import { XMLParser } from "fast-xml-parser";

export type RSSFeed = {
  channel: {
    title: string;
    link: string;
    description: string;
    item: RSSItem[];
  }
}

export type RSSItem = {
  title: string;
  link: string;
  description: string;
  pubDate: string;
}

export function isRSSFeed(value: any): value is RSSFeed {
  if (value.channel === undefined) {
    return false;
  }
  if (
    typeof value.channel.title !== "string" ||
    typeof value.channel.link !== "string" ||
    typeof value.channel.description !== "string" ||
    !Array.isArray(value.channel.item)
  ) {
    return false;
  }

  return true;
}

export function isRSSItem(value: any): value is RSSItem {
  if (
    typeof value.title !== "string" ||
    typeof value.link !== "string" ||
    typeof value.description !== "string" ||
    typeof value.pubDate !== "string"
  ) {
    return false;
  }

  return true;
}

export async function fetchFeed(feedURL: string) {
  const response = await fetch(feedURL, {
    method: "GET",
    mode: "cors",
    headers: {
      "User-Agent": "gator"
    }
  });

  const feed = await response.text();

  const parser = new XMLParser();
  let feedObj = parser.parse(feed);

  if (!isRSSFeed(feedObj?.rss)) {
    throw new Error("Not an RSS feed");
  }
  feedObj = feedObj.rss;
  const items: RSSItem[] = [];
  for (const item of feedObj.channel.item) {
    if (isRSSItem(item)) {
      items.push(item);
    }
  }
  feedObj.channel.item = items;

  return feedObj;
}

